downcast() should return a smart pointer when passed in a smart pointer

cdumez opened the pull request:
#### 245867d0781caf78b6b400c2f97386488b38254c
<pre>
downcast() should return a smart pointer when passed in a smart pointer
<a href="https://bugs.webkit.org/show_bug.cgi?id=261415">https://bugs.webkit.org/show_bug.cgi?id=261415</a>

Reviewed by NOBODY (OOPS!).

downcast() should return a smart pointer when passed in a smart pointer.
Otherwise, it forces us to use the less-safe static_pointer_cast() /
static_reference_cast() to cast without causing ref counting churn.

This will help with our increased adoption of smart pointers in WebKit.

* Source/WTF/wtf/Ref.h:
(WTF::is):
(WTF::downcast):
(WTF::dynamicDowncast):
* Source/WTF/wtf/RefPtr.h:
(WTF::is):
(WTF::downcast):
* Source/WTF/wtf/TypeCasts.h:
(WTF::is):
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
* Source/WebCore/bindings/js/JSElementCustom.cpp:
(WebCore::createNewElementWrapper):
* Source/WebCore/bindings/js/JSNodeCustom.cpp:
(WebCore::createWrapperInline):
* Source/WebCore/bindings/js/JSWebXRSpaceCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
(WebCore::nameFromCSSValue):
(WebCore::additiveSymbolsFromCSSValue):
(WebCore::padFromCSSValue):
(WebCore::symbolsFromCSSValue):
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::setSelectorText):
(WebCore::CSSStyleRule::insertRule):
(WebCore::CSSStyleRule::deleteRule):
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::traverseRulesInVector):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeFontFeatureValuesRule):
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
(WebCore::CSSNumericValue::toSum):
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::set):
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::getImageData):
* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::HTMLSlotElement::assignedElements const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::getImageData const):
* Source/WebCore/html/track/WebVTTParser.cpp:
(WebCore::WebVTTParser::checkAndStoreStyleSheet):
* Source/WebCore/loader/SubstituteResource.h:
(WebCore::SubstituteResource::data const):
* Source/WebCore/page/DOMTimer.cpp:
(WebCore::DOMTimerFireState::DOMTimerFireState):
(WebCore::DOMTimerFireState::contextDocument const):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragment):
* Source/WebCore/page/PerformanceUserTiming.cpp:
(WebCore::PerformanceUserTiming::mark):
* Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
(WebCore::ScrollingStateTree::commit):
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::enqueueSample):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):
* Source/WebCore/style/StyleInvalidator.cpp:
(WebCore::Style::shouldDirtyAllStyle):
* Source/WebCore/style/UserAgentStyle.cpp:
(WebCore::Style::UserAgentStyle::addToDefaultStyle):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performImmediateActionHitTestAtLocation):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/245867d0781caf78b6b400c2f97386488b38254c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18054 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18382 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18949 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19888 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16902 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18252 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21679 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18540 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18886 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18272 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18527 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15715 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20765 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15753 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16468 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23003 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/15646 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16770 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16638 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20876 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/17342 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17203 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14585 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/21404 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16300 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5246 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4301 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20662 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/22639 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17052 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5093 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/Ref.h:
```diff
@@ -265,29 +265,52 @@ inline Ref<T, U> adoptRef(T& reference)
 
 template<typename Target, typename Source, typename PtrTraits>
-inline Target& downcast(Ref<Source, PtrTraits>& source)
+inline Ref<Target> downcast(Ref<Source, PtrTraits>& source)
```
Not sure why this pre-existing overload is needed. `dynamicDowncast` doesn't have it.

darinadler commented on Source/WTF/wtf/Ref.h:
```diff
@@ -265,29 +265,52 @@ inline Ref<T, U> adoptRef(T& reference)
+    static_assert(!std::is_same_v<Source, Target>, "Unnecessary cast to same type");
+    static_assert(std::is_base_of_v<Source, Target>, "Should be a downcast");
+    ASSERT_WITH_SECURITY_IMPLICATION(is<Target>(source));
+    return static_reference_cast<Target>(WTFMove(source));
```
Since this is a template, it's a universal reference and not an rvalue reference and so we should be using `std::forward` instead of `WTFMove` and may not need the `const&` overload.

darinadler commented on Source/WTF/wtf/Ref.h:
```diff
@@ -265,29 +265,52 @@ inline Ref<T, U> adoptRef(T& reference)
+    static_assert(std::is_base_of_v<Source, Target>, "Should be a downcast");
+    if (!is<Target>(source))
+        return nullptr;
+    return static_reference_cast<Target>(WTFMove(source));
```
Since this is a template, it's a universal reference and not an rvalue reference and so we should be using `std::forward` instead of `WTFMove` and may not need the `const&` overload.

darinadler commented on Source/WTF/wtf/RefPtr.h:
```diff
@@ -254,25 +254,28 @@ struct IsSmartPtr<RefPtr<T, U, V>> {
+    static_assert(!std::is_same_v<Source, Target>, "Unnecessary cast to same type");
+    static_assert(std::is_base_of_v<Source, Target>, "Should be a downcast");
+    ASSERT_WITH_SECURITY_IMPLICATION(!source || is<Target>(*source));
+    return static_reference_cast<Target>(WTFMove(source));
```
Since this is a template, it's a universal reference and not an rvalue reference and so we should be using `std::forward` instead of `WTFMove` and may not need the `const&` overload.

darinadler commented on Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
```diff
@@ -87,8 +87,7 @@ static CSSCounterStyleDescriptors::Name nameFromCSSValue(Ref<CSSValue> value)
 
-    auto& primitiveValue = downcast<CSSPrimitiveValue>(value);
-    return makeAtomString(primitiveValue.stringValue());
+    return makeAtomString(downcast<CSSPrimitiveValue>(value.get()).stringValue());
```
This could use `WTFMove` instead of `get()`. Not sure which is better.

darinadler commented on Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
```diff
@@ -102,7 +101,7 @@ static CSSCounterStyleDescriptors::AdditiveSymbols additiveSymbolsFromStylePrope
 {
     CSSCounterStyleDescriptors::AdditiveSymbols result;
-    for (auto& additiveSymbol : downcast<CSSValueList>(value)) {
+    for (auto& additiveSymbol : downcast<CSSValueList>(value.get())) {
```
This could use `WTFMove` instead of `get()`. Not sure which is better.

cdumez replied:
But then I need to call `.get()` on the value returned by `WTFMove()` to make the range loop work, so it is not necessarily better.

darinadler commented on Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
```diff
@@ -121,7 +120,7 @@ static CSSCounterStyleDescriptors::Pad padFromStyleProperties(const StylePropert
 CSSCounterStyleDescriptors::Pad padFromCSSValue(Ref<CSSValue> value)
 {
-    auto& list = downcast<CSSValueList>(value);
+    auto& list = downcast<CSSValueList>(value.get());
```
This could use `WTFMove` instead of `get()`. Not sure which is better.

darinadler commented on Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
```diff
@@ -159,7 +158,7 @@ static Vector<CSSCounterStyleDescriptors::Symbol> symbolsFromStyleProperties(con
 {
     Vector<CSSCounterStyleDescriptors::Symbol> result;
-    for (auto& symbolValue : downcast<CSSValueList>(value)) {
+    for (auto& symbolValue : downcast<CSSValueList>(value.get())) {
```
This could use `WTFMove` instead of `get()`. Not sure which is better.

darinadler commented on Source/WebCore/css/CSSStyleRule.cpp:
```diff
@@ -125,7 +125,7 @@ void CSSStyleRule::setSelectorText(const String& selectorText)
 
     if (m_styleRule->isStyleRuleWithNesting())
-        downcast<StyleRuleWithNesting>(m_styleRule).wrapperAdoptOriginalSelectorList(WTFMove(*selectorList));
+        downcast<StyleRuleWithNesting>(m_styleRule)->wrapperAdoptOriginalSelectorList(WTFMove(*selectorList));
```
It's kind of strange to always create a new `Ref` here as part of downcast giving a `Ref` when passed a `Ref` yet in other places use `get()` so we get a raw pointer or reference. What is the distinction?

cdumez replied:
`wrapperAdoptOriginalSelectorList()` seemed like a non-trivial function so it is best to call it on a smart pointer variable we have on the stack. Even though `m_styleRule` is a smart pointer, it could in theory get destroyed if `this` gets destroyed. For this reason, I chose to make a explicit Ref here.

rreno commented on Source/WTF/wtf/RefPtr.h:
```diff
@@ -254,25 +254,28 @@ struct IsSmartPtr<RefPtr<T, U, V>> {
+    static_assert(!std::is_same_v<Source, Target>, "Unnecessary cast to same type");
+    static_assert(std::is_base_of_v<Source, Target>, "Should be a downcast");
+    ASSERT_WITH_SECURITY_IMPLICATION(!source || is<Target>(*source));
+    return static_reference_cast<Target>(WTFMove(source));
```
Should this be `static_pointer_cast` instead of `static_reference_cast`?

cdumez replied:
Yes, will fix.

darinadler commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -1209,7 +1209,7 @@ static bool sampleBufferRenderersSupportKeySession()
         return;
 
-    Ref<MediaSampleAVFObjC> sampleAVFObjC = static_reference_cast<MediaSampleAVFObjC>(WTFMove(sample));
+    Ref<MediaSampleAVFObjC> sampleAVFObjC = downcast<MediaSampleAVFObjC>(WTFMove(sample));
```
Just `Ref` or `auto`?

rreno commented on Source/WebCore/page/PerformanceUserTiming.cpp:
```diff
@@ -100,7 +100,7 @@ ExceptionOr<Ref<PerformanceMark>> PerformanceUserTiming::mark(JSC::JSGlobalObjec
         timestamp = m_performance.monotonicTimeFromRelativeTime(*markOptions->startTime);
 
-    InspectorInstrumentation::performanceMark(context.get(), markName, timestamp, is<Document>(context) ? downcast<Document>(context).frame() : nullptr);
+    InspectorInstrumentation::performanceMark(context.get(), markName, timestamp, is<Document>(context) ? downcast<Document>(context.get()).frame() : nullptr);
```
Just picking this as an example, there's other examples of this pattern in the patch.

Are `foo.get()` (or `foo.ptr()`) necessary? I think if you just pass `context` here you'll get the const ref overload of downcast which will call `get` for you, right?

It would be a nice API for downcast if we don't have to think about getting the raw pointer or raw reference out of the smart pointer when downcasting which I think is how it would work as written.

cdumez replied:
`downcast<Document>(context).frame()` wouldn't build because `context` is a smart pointer and so downcast would return a smart pointer to. I could have used `downcast<Document>(context)->frame()` to make it build but that would create a little bit of unnecessary ref-counting churn. As a result, I decided to go with `downcast<Document>(context.get()).frame()` to avoid the churn (minor optimization).

cdumez replied:
Or were you talking about the first parameter (`context.get()`) which I didn't change in my patch?

rreno replied:
I was talking about the downcast parameter, I should have picked a better example where there wasn't an ambiguity.

That makes sense, you specifically wanted the raw reference in this case and likely in all the other cases where you use `.get` when calling `downcast`.

rreno reviewed (commented):
LGTM with a question. You'll need an official r+ from someone with reviewer status though.

darinadler reviewed (approved):
Seems to me that this should take a `Ref` or `RefPtr` so we can transfer ownership when appropriate. Using `const Ref&` requires reference count churn. Using `Ref&&` would require extra work for each caller that doesn't transfer ownership. But just `Ref` lets us move if appropriate or copy if not. Or perhaps we can accomplish this with overloading.

cdumez committed:
downcast() should return a smart pointer when passed in a smart pointer
https://bugs.webkit.org/show_bug.cgi?id=261415

Reviewed by Darin Adler.

downcast() should return a smart pointer when passed in a smart pointer.
Otherwise, it forces us to use the less-safe static_pointer_cast() /
static_reference_cast() to cast without causing ref counting churn.

This will help with our increased adoption of smart pointers in WebKit.

* Source/WTF/wtf/Ref.h:
(WTF::is):
(WTF::downcast):
(WTF::dynamicDowncast):
* Source/WTF/wtf/RefPtr.h:
(WTF::is):
(WTF::downcast):
* Source/WTF/wtf/TypeCasts.h:
(WTF::is):
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
* Source/WebCore/bindings/js/JSElementCustom.cpp:
(WebCore::createNewElementWrapper):
* Source/WebCore/bindings/js/JSNodeCustom.cpp:
(WebCore::createWrapperInline):
* Source/WebCore/bindings/js/JSWebXRSpaceCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
(WebCore::nameFromCSSValue):
(WebCore::additiveSymbolsFromCSSValue):
(WebCore::padFromCSSValue):
(WebCore::symbolsFromCSSValue):
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::setSelectorText):
(WebCore::CSSStyleRule::insertRule):
(WebCore::CSSStyleRule::deleteRule):
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::traverseRulesInVector):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeFontFeatureValuesRule):
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
(WebCore::CSSNumericValue::toSum):
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::set):
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::getImageData):
* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::HTMLSlotElement::assignedElements const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::getImageData const):
* Source/WebCore/html/track/WebVTTParser.cpp:
(WebCore::WebVTTParser::checkAndStoreStyleSheet):
* Source/WebCore/loader/SubstituteResource.h:
(WebCore::SubstituteResource::data const):
* Source/WebCore/page/DOMTimer.cpp:
(WebCore::DOMTimerFireState::DOMTimerFireState):
(WebCore::DOMTimerFireState::contextDocument const):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragment):
* Source/WebCore/page/PerformanceUserTiming.cpp:
(WebCore::PerformanceUserTiming::mark):
* Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
(WebCore::ScrollingStateTree::commit):
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::enqueueSample):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):
* Source/WebCore/style/StyleInvalidator.cpp:
(WebCore::Style::shouldDirtyAllStyle):
* Source/WebCore/style/UserAgentStyle.cpp:
(WebCore::Style::UserAgentStyle::addToDefaultStyle):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performImmediateActionHitTestAtLocation):

Canonical link: https://commits.webkit.org/267923@main
